processing/src/main/java/org/apache/druid/query/scan/ScanQueryOrderByMergeRowIterator.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.druid.utils.CollectionUtils;
+
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ScanQueryOrderByMergeRowIterator extends ScanQueryLimitRowIterator
+{
+
+  public ScanQueryOrderByMergeRowIterator(
+      QueryRunner<ScanResultValue> baseRunner,
+      QueryPlus<ScanResultValue> queryPlus,
+      ResponseContext responseContext
+  )
+  {
+    super(baseRunner, queryPlus, responseContext);
+    if (ScanQuery.ResultFormat.RESULT_FORMAT_VALUE_VECTOR.equals(resultFormat)) {
+      throw new UOE(ScanQuery.ResultFormat.RESULT_FORMAT_VALUE_VECTOR + " is not supported yet");
+    }
+  }
+
+  @Override
+  public boolean hasNext()
+  {
+    return !yielder.isDone();
+  }
+
+  @Override
+  public ScanResultValue next()
+  {
+
+    final int scanRowsLimit;
+    if (query.getScanRowsLimit() > Integer.MAX_VALUE) {
+      scanRowsLimit = Integer.MAX_VALUE;
+    } else {
+      scanRowsLimit = Math.toIntExact(query.getScanRowsLimit());
+    }
+
+    List<String> orderByDirection = query.getOrderBys()
+                                         .stream()
+                                         .map(orderBy -> orderBy.getOrder().toString())
+                                         .collect(Collectors.toList());
+    List<Integer> sortColumnIdxs = query.getSortColumnIdxs();
+
+    Ordering<Comparable>[] orderings = new Ordering[orderByDirection.size()];
+    for (int i = 0; i < orderByDirection.size(); i++) {
+      orderings[i] = ScanQuery.Order.ASCENDING.equals(ScanQuery.Order.fromString(orderByDirection.get(i)))
+                     ? Comparators.naturalNullsFirst()
+                     : Comparators.<Comparable>naturalNullsFirst().reverse();
+    }
+
+    Comparator<List<Object>> comparator = (o1, o2) -> {
+      for (int i = 0; i < sortColumnIdxs.size(); i++) {
+        int compare = orderings[i].compare(
+            (Comparable) o1.get(sortColumnIdxs.get(i)),
+            (Comparable) o2.get(sortColumnIdxs.get(i))
+        );
+        if (compare != 0) {
+          return compare;
+        }
+      }
+      return 0;
+    };
+    Sorter<Object> sorter = new QueueBasedSorter<Object>(scanRowsLimit, comparator);
+    List<String> columns = new ArrayList<>();
+    while (!yielder.isDone()) {
+      ScanResultValue srv = yielder.get();
+      columns = columns.isEmpty() ? srv.getColumns() : columns;
+      List<List<Object>> events = (List<List<Object>>) srv.getEvents();
+      for (Object event : events) {
+        if (event instanceof LinkedHashMap) {
+          sorter.add(Arrays.asList(((LinkedHashMap) event).values().toArray()));
+        } else {
+          sorter.add((List<Object>) event);
+        }
+      }
+      yielder = yielder.next(null);
+      count++;
+    }
+    final List<List<Object>> sortedElements = new ArrayList<>(sorter.size());
```

Comment:
As in the earlier note: a scan result value has a maximum batch size. We have to deliver results in batches no larger than that size.

Since "deliver of limited size batches from a sorter" is common between the merge and scan engine steps, it would be great if we could factor out a common implementation. That way, I don't have to comment on basically the same code twice. DRY and all that.